editor/editor_group_settings.cpp

```diff
@@ -0,0 +1,745 @@
+            undo_redo->add_undo_method(ProjectSettings::get_singleton(), "set_order", swap_name, swap_order);
+
+            undo_redo->add_do_method(this, "update_tree");
+            undo_redo->add_undo_method(this, "update_tree");
+
+            undo_redo->add_do_method(this, "emit_signal", group_changed);
+            undo_redo->add_undo_method(this, "emit_signal", group_changed);
+
+            undo_redo->commit_action();
+            break;
+        } case BUTTON_DELETE:
+            remove_group(ti->get_text(0));
+            break;
+    }
+}
+
+void EditorGroupSettings::_get_all_scenes(EditorFileSystemDirectory *p_dir, Set<String> &r_list) {
+
+    for (int i = 0; i < p_dir->get_file_count(); i++) {
+        if (p_dir->get_file_type(i) == "PackedScene")
+            r_list.insert(p_dir->get_file_path(i));
+    }
+
+    for (int i = 0; i < p_dir->get_subdir_count(); i++) {
+        _get_all_scenes(p_dir->get_subdir(i), r_list);
+    }
+}
+
+void EditorGroupSettings::_fix_remove_group(const String &p_name) {
+
+    // apply fix to all loaded scenes
+    Node *es = EditorNode::get_singleton()->get_edited_scene();
+
+    if (es != NULL && _fix_node_remove_group(es, p_name))
+    {
+        EditorNode::get_singleton()->get_scene_tree_dock()->get_tree_editor()->update_tree();
+    }
+
+    // apply fix to all other scenes
+    Node *node;
+    Set<String> scenes;
+    _get_all_scenes(EditorFileSystem::get_singleton()->get_filesystem(), scenes);
+    for (Set<String> ::Element *E = scenes.front(); E; E = E->next()) {
+
+        String path = E->get();
+
+        if (!EditorNode::get_singleton()->is_scene_in_use(path)) {
+
+            Ref<PackedScene> data = ResourceLoader::load(path);
+
+            PackedScene::GenEditState ges = PackedScene::GEN_EDIT_STATE_INSTANCE;
+            node = data->instance(ges);
+
+            if (_fix_node_remove_group(node, p_name)) {
+                data->pack(node);
+                ResourceSaver::save(path, data);
+            }
+            memdelete(node);
+            node = NULL;
+        }
+    }
+}
+
+void EditorGroupSettings::_fix_rename_group(const String &p_old_name, const String &p_new_name) {
+
+    // apply fix to all loaded scenes
+    Node *es = EditorNode::get_singleton()->get_edited_scene();
+    if (es != NULL && _fix_node_rename_group(es, p_old_name, p_new_name))
+    {
+        EditorNode::get_singleton()->get_scene_tree_dock()->get_tree_editor()->update_tree();
+    }
+
+    // apply fix to all other scenes
+    Node *node;
+    Set<String> scenes;
+    _get_all_scenes(EditorFileSystem::get_singleton()->get_filesystem(), scenes);
+    for (Set<String> ::Element *E = scenes.front(); E; E = E->next()) {
+
+        String path = E->get();
+
+        if (!EditorNode::get_singleton()->is_scene_in_use(path)) {
+
+            Ref<PackedScene> data = ResourceLoader::load(path);
+
+            PackedScene::GenEditState ges = PackedScene::GEN_EDIT_STATE_INSTANCE;
```

Comment:
Every scenario where you are instancing the PackedScene should be converted into something that calls [SceneState.get_node_groups](http://docs.godotengine.org/en/latest/classes/class_scenestate.html#class-scenestate-get-node-groups). It should never be actually instancing all of the scenes in the project (that could be colossally slow on larger projects).

Reply:
Hmmm I get you.. but then how can I change and remove nodes from groups. I should only instance the ones that are needed is that what you are saying? I'll get to it now!

Reply:
I can't actually edit node groups using SceneState only. I have to instantiate it, unfortunately